shai/src/main/scala/at/forsyte/apalache/shai/TransExplorerService.scala

```diff
@@ -108,8 +124,11 @@ class TransExplorerService(connections: Ref[Map[UUID, Conn]], parserSemaphore: S
    */
   def openConnection(req: ConnectRequest): Result[Connection] = for {
     id <- ZIO.effectTotal(UUID.randomUUID())
+    conn = Conn(id)
     _ <- setConnection(Conn(id))
-  } yield Connection(id.toString())
+    connection = Connection(id.toString())
+    _ <- Log.info(Some(connection), "New connection created")
```

Comment:
Yeah, that's not possible. These values need to return ZIO vaules, which are "promises" or "tasks" that can be executed concurrently. That's the purpose of our wrappers.

Reply:
IUC and these indeed return promises, does it mean that all the logger calls run concurrently? I'm not even sure we'd want that.

In general, I'm still confused when and where ZIO exactly introduces concurrency. I tried reading the docs, but haven't really found an explanation. Can you maybe point us somewhere?

My fundamental confusion is the following:
Even if we wrap the `logger.info` call inside an effect, it will internally make some unwrapped calls. So at some level we break out of the ZIO effect framework no matter what. Then why do we try to lift it into our code here?

It'd seem enough to me to introduce concurrency/promises at one level, which is when we dispatch calls to the gRPC server. That is, each gRPC connection is handled concurrently, but anything inside a connection can run sequentially.

Reply:
I mean, all the logger calls run concurrently along with whatever other operations performed by the server (so it might be server one request while logging the response for another) but the logging calls get sequenced within the particular handler invocations. I’ve tried to cover the mechanics of this in the README. I’d there’s some missing exposition there that could help clarify, I’d be happy to add it.

> Even if we wrap the logger.info call inside an effect, it will internally make some unwrapped calls (i.e., outside of ZIO). So at some level we break out of the ZIO effect framework no matter what. Then why do we try to lift it into our code here?

The basic idea is: We first build a data structure representing the concurrent computations we want to run, then the scheduler executes the computations, being sure to manage all the concurrency and sequencing.

The rule of thumb is: any effects must be lifted into the data structure before being executed. That’s all a reader should need to know to use and read the code that uses ZIO here. But I can try to provide more insight.

When we lift effects into ZIO values we represent an effect that should eventually be enacted, but nothing happens at that point. It’s just data. When ZIO’s task scheduler runs, it will begin executing the events in ways consistent with the order described in the data structure we’ve built (running things concurrently or in sequence, as appropriate). At that point, it can execute effects, and they leave ZIO managed land. But if we execute the effect when we are building the data structure rather than when the scheduler deems appropriate, we lose the benefit of delegating the concurrency execution to the scheduler: most notably, we lose the ability to ensure the effect is sequenced when we want it to be. 

> It'd seem enough to me to introduce concurrency/promises at one level, which is when we dispatch calls to the gRPC server. That is, there should be concurrency between handling different gRPC connections, but anything inside a connection can run sequentially.

That’s correct. The way we mode sequential effects here is within the for comprehensions (or by use of the combinators like `andThen`).